File: BackEnd/src/main/java/ma/ensate/rh/ensate_rh_backend/model/Diplome.java
package ma.ensate.rh.ensate_rh_backend.model;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.LocalDate;

@Entity
@Table(name = "diplome")
@Data
@NoArgsConstructor
@AllArgsConstructor
public class Diplome {
    
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;
    
    @Column(name = "intitule", length = 100, nullable = false)
    private String intitule;
    
    @Column(name = "specialite", length = 100)
    private String specialite;
    
    @Convert(converter = NiveauConverter.class)
    @Column(name = "niveau", length = 20)
    private Niveau niveau;
    
    @Column(name = "etablissement", length = 150)
    private String etablissement;
    
    @Column(name = "date_obtention")
    private LocalDate dateObtention;
    
    @Column(name = "fichier_preuve", length = 255)
    private String fichierPreuve;
    
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "personnel_id", nullable = false)
    private Personnel personnel;
    
    public enum Niveau {
        Bac, 
        Bac_Plus_2, 
        Licence, 
        Master, 
        Ingénieur, 
        Doctorat
    }
}

File: FrontEnd/app/src/main/java/com/ensa/gestionpersonnel/domain/model/Diplome.kt
package com.ensa.gestionpersonnel.domain.model

import java.util.Date

/**
 * Énumération des niveaux de diplômes
 */
enum class NiveauDiplome {
    Bac,
    Bac_Plus_2,
    Licence,
    Master,
    Ingénieur,
    Doctorat
}

/**
 * Modèle représentant un diplôme d'un personnel
 */
data class Diplome(
    val id: Long = 0L,
    val intitule: String,
    val personnelId: Long,
    val specialite: String,
    val niveau: NiveauDiplome,
    val etablissement: String,
    val dateObtention: Date?,
    val fichierPreuve: String? = null,
    val personnelNom: String? = null,
    val personnelPrenom: String? = null
) {

    /**
     * Vérifie si le diplôme est valide
     * - L'intitulé ne doit pas être vide
     * - La date d'obtention ne doit pas être dans le futur
     */
    fun estValide(): Boolean {
        return intitule.isNotBlank() &&
                specialite.isNotBlank() &&
                etablissement.isNotBlank() &&
                (dateObtention?.before(Date()) ?: true)
    }

    /**
     * Retourne le nom complet du diplôme
     */
    fun getNomComplet(): String {
        return "$intitule - $specialite ($niveau)"
    }

    /**
     * Calcule l'ancienneté du diplôme en années
     */
    fun getAncienneteAnnees(): Int {
        if (dateObtention == null) return 0
        val diff = Date().time - dateObtention.time
        return (diff / (1000L * 60 * 60 * 24 * 365)).toInt()
    }

    /**
     * Vérifie si un fichier justificatif est attaché
     */
    fun hasPreuve(): Boolean = !fichierPreuve.isNullOrBlank()

    /**
     * Retourne le niveau numérique (utile pour les comparaisons)
     */
    fun getNiveauNumerique(): Int {
        return when (niveau) {
            NiveauDiplome.Bac -> 1
            NiveauDiplome.Bac_Plus_2 -> 2
            NiveauDiplome.Licence -> 3
            NiveauDiplome.Master -> 4
            NiveauDiplome.Ingénieur -> 4
            NiveauDiplome.Doctorat -> 5
        }
    }
}